### PR TITLE
Fix label clipping by switching to layer-by-layer rendering

### LINKED ISSUE #757: Labels slipping/cropping at tile bounds
As we pan in and out, labels tuck under other layers at certain fractional zoom levels. So far, I've only seen this happen with city labels.

EDIT: happens with all symbols.

### PR DESCRIPTION
**Pull request by lucaswoj**
cc @mourner @jfirebaugh @ansis @kkaefer 

fixes #757 

# Remaining Tasks

 - [ ] remove [reverse iteration over features (how does mapbox-gl-native do this?)](https://github.com/mapbox/mapbox-gl-js/pull/1727/files#r47952469)

### CODE REVIEW

**Comment by lucaswoj** [comment]
There's still a chunk of work to be done here but rendering is already *so* much smoother than `master`

**Comment by mourner** [comment]
> command npm run preinstall && npm update && npm run postinstall && npm run prepublish took more than 120 minutes to run

Oh wow, wtf.

**Comment by yhahn** [comment]
:open_mouth:

**Comment by mourner** [comment]
> I spent a long time trying to do the rebase without squashing but it was really difficult due to broken intermediate states.

When I did a similar thing for the earcut branch, it was much easier to manually look through the old PR bit by bit and enter the changes manually on a fresh master-based branch. Rebasing is hell on very outdated branches, and even if you end up making it work, there's still a lot of broken stuff usually.

**Comment by lucaswoj** [comment]
It looks like symbol render order (namely, overlapping symbols within the same layer) is going different with layer-by-layer rendering. This might be acceptable for our users but is not acceptable for `mapbox-gl-test-suite`.

**Comment by jfirebaugh** [comment]
Symbols should be z-ordered by y coordinate. AFAIK, both gl-js and native (which already uses layer-by-layer rendering) agree on this currently; if this PR changes the ordering in gl-js, it's probably a bug.

**Comment by lucaswoj** [comment]
Ah! Ok. I didn't know there was a deterministic symbol rendering order. I'll keep working on that one. 

Some other test-suite failures appear to be `headless-gl` specific. They are not replicable in the browser. This one occurs whether we pull in earcut or not.

# Test Suite

screen shot 2015-11-16 at 2 57 10 pm

# Browser

screen shot 2015-11-16 at 2 57 27 pm

**Comment by lucaswoj** [comment]
I'm pleased to report that the fill rendering problems are resolved now (along with a number of other problems). This is shaping up nicely!

**Comment by lucaswoj** [comment]
Interesting clue on the disappearing fills: rendering all fills during the "translucent" pass fixes almost all of the fill test cases.

**Comment by lucaswoj** [comment]
I think I'm onto something -- not rendering `background` layers fixes most of the rendering problems in test-suite.

**Comment by lucaswoj** [comment]
While I still can't reproduce the exact rendering errors that I'm seeing in headless-gl, I have stumbled across a test case that produces seemingly-related wrong behavior in the browser. These two styles render exactly the same, even though their layer order is reversed.

```js
{
    "version": 8,
    "sources": {
      "mapbox": {
          "type": "vector",
          "url": "mapbox://mapbox.mapbox-streets-v6"
      }
    },
    "layers": [{
        "id": "background",
        "type": "background",
        "paint": {
            "background-color": "red"
        }
    },
    {
        "id": "water",
        "type": "fill",
        "source": "mapbox",
        "source-layer": "water",
        "paint": {
            "fill-color": "blue",
            "fill-antialias": false
        }
    }]
}
```

```js
{
    "version": 8,
    "sources": {
      "mapbox": {
          "type": "vector",
          "url": "mapbox://mapbox.mapbox-streets-v6"
      }
    },
    "layers": [{
        "id": "water",
        "type": "fill",
        "source": "mapbox",
        "source-layer": "water",
        "paint": {
            "fill-color": "blue",
            "fill-antialias": false
        }, {
        "id": "background",
        "type": "background",
        "paint": {
            "background-color": "red"
        }
    }]
}
```

**Comment by lucaswoj** [comment]
only 25 failing test-suite tests :exclamation:

**Comment by lucaswoj** [comment]
Before optimization, fps perf seems ballpark ok

```
duration: 3000
44 fps https://api.tiles.mapbox.com/mapbox-gl-js/v0.10.0/mapbox-gl.js
51 fps https://api.tiles.mapbox.com/mapbox-gl-js/v0.12.1/mapbox-gl.js
47.7 fps /dist/mapbox-gl.js
42.1 fps https://api.tiles.mapbox.com/mapbox-gl-js/v0.10.0/mapbox-gl.js
52.1 fps https://api.tiles.mapbox.com/mapbox-gl-js/v0.12.1/mapbox-gl.js
48.4 fps /dist/mapbox-gl.js
```

File: js/data/bucket.js [hunk 1]
```diff
@@ -81,7 +81,7 @@ function Bucket(options) {
  * @private
  */
 Bucket.prototype.addFeatures = function() {
-    for (var i = 0; i < this.features.length; i++) {
+    for (var i = this.features.length - 1; i >= 0; i--) {
```
Comment (lucaswoj) [inline js/data/bucket.js]: This isn't quite right. We only want to reverse the feature order during translucent passes.

**Comment by lucaswoj** [comment]
This is ready for some serious :eyes: @mourner @jfirebaugh @ansis @tmcw @kkaefer.

**Comment by lucaswoj** [comment]
If you want to follow the full commit history, it begins at #1197

File: js/data/bucket.js (hunk 1 above)
Comment (jfirebaugh) [inline js/data/bucket.js]: Why is the iteration order reversed here (and in `LineBucket`)?

Comment (lucaswoj) [inline js/data/bucket.js]: Because we are using the `gl.ONE_MINUS_DST_ALPHA` blending function, features are effectively rendered in reverse order.

Comment (jfirebaugh) [inline js/data/bucket.js]: Hmm. Native does not work this way. [It iterates forward.](https://github.com/mapbox/mapbox-gl-native/blob/d5fb535058d4d97567b08064e0203f45c2c8370f/src/mbgl/style/style_bucket_parameters.cpp#L8-L8)

If this is affecting rendering there must be some other difference between native and JS that is reversed.

File: js/render/draw_circle.js [hunk 2]
```diff
@@ -4,27 +4,19 @@ var browser = require('../util/browser.js');
 
 module.exports = drawCircles;
 
-function drawCircles(painter, layer, posMatrix, tile) {
-    // short-circuit if tile is empty
-    if (!tile.buffers) return;
-
-    posMatrix = painter.translateMatrix(posMatrix, tile, layer.paint['circle-translate'], layer.paint['circle-translate-anchor']);
+function drawCircles(painter, source, layer, coords) {
+    var gl = painter.gl;
```
Comment (jfirebaugh) [inline js/render/draw_circle.js]: Native [aborts early if it's not the translucent pass.](https://github.com/mapbox/mapbox-gl-native/blob/45c7b472adb78600ab9f9c8f82082d6b0af21ba2/src/mbgl/renderer/painter_circle.cpp#L18-L18) Do we need that here?

File: js/data/bucket.js (hunk 1 above)
Comment (lucaswoj) [inline js/data/bucket.js]: I will spend some more time trying to understand this, I don't have an immediate explanation. This is what I had to do to make test-suite pass :grimacing:

File: js/render/draw_circle.js (hunk 2 above)
Comment (lucaswoj) [inline js/render/draw_circle.js]: Yup. Added.

File: js/data/bucket.js (hunk 1 above)
Comment (lucaswoj) [inline js/data/bucket.js]: I would like to note that this reversal might lead to some rendering bugs. While rendering order is reversed during translucent passes, it is in the expected order during opaque passes.

Comment (ansis) [inline js/data/bucket.js]: The reverse iteration here makes the rendered order of features more similar to the previous rendered order, but that previous order was arbitrary. I think it should be safe to remove this and update the render tests.

**Comment by ansis** [comment]
squashed, rebased and merged in https://github.com/mapbox/mapbox-gl-js/pull/1886

**Pull request closed without merging**